test/parallel/test-http-agent.js

```diff
@@ -20,41 +20,55 @@
 // USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 'use strict';
-require('../common');
+const common = require('../common');
 const assert = require('assert');
 const http = require('http');
 
-const server = http.Server(function(req, res) {
-  res.writeHead(200);
-  res.end('hello world\n');
-});
-
 let responses = 0;
 const N = 4;
 const M = 4;
 
-server.listen(0, function() {
-  const port = this.address().port;
-  for (let i = 0; i < N; i++) {
-    setTimeout(function() {
-      for (let j = 0; j < M; j++) {
-        http.get({ port: port, path: '/' }, function(res) {
-          console.log('%d %d', responses, res.statusCode);
-          if (++responses === N * M) {
-            console.error('Received all responses, closing server');
-            server.close();
-          }
-          res.resume();
-        }).on('error', function(e) {
-          console.log('Error!', e);
-          process.exit(1);
-        });
+const server = http.Server(common.mustCall(function(req, res) {
+  res.writeHead(200);
+  res.end('hello world\n');
+}, (N * M) + 1));
+
+function makeRequests(outCount, inCount) {
+  return new Promise((resolve) => {
+    const onRes = common.mustCall(function(res) {
+      console.log('%d %d', responses, res.statusCode);
+      if (++responses === outCount * inCount) {
+        server.close();
+        resolve();
       }
-    }, i);
-  }
-});
+      res.resume();
+    }, outCount * inCount);
 
+    server.listen(0, () => {
+      const port = server.address().port;
+      for (let i = 0; i < outCount; i++) {
+        setTimeout(() => {
+          for (let j = 0; j < inCount; j++) {
+            http
+              .get({ port: port, path: '/' }, onRes)
+              .on('error', (e) => assert.fail(e));
+          }
+        }, i);
+      }
+    });
+  });
+}
 
-process.on('exit', function() {
-  assert.strictEqual(N * M, responses);
+makeRequests(N, M).then(() => {
+  // should not explode if can not create sockets
+  // Ref: https://github.com/nodejs/node/issues/13045
+  // Ref: https://github.com/nodejs/node/issues/13831
+  http.Agent.prototype.createConnection = function createConnection(_, cb) {
+    process.nextTick(cb, new Error('nothing'));
+  };
+  return makeRequests(N, M);
+}).catch((e) => {
+  // This is currently the way to fail a test with a Promise
+  console.error(e);
+  process.exit(1);
```

Comment:
Should we throw the `fatalError()` code in `lib/async_hooks.js` into `test/common` (https://github.com/nodejs/node/blob/v8.1.3/lib/async_hooks.js#L54-L68)? This way there's a standardized way of failing. e.g.
```js
const { promiseFatalError } = require('../common'):

Promise.reject(42).catch(promiseFatalError);
```